### PR TITLE
Add the possibility to deprecate attributes and nodes on Node

### PR DESCRIPTION
**Pull request by fabpot**

### CODE REVIEW

**Comment by stof** [comment]
How does deprecating a node interacts with node visitors ?

**Comment by fabpot** [comment]
> How does deprecating a node interacts with node visitors ?

That's the main use case for me: you get deprecations as node visitors call `setAttribute()/setNode()`.

**Comment by stof** [comment]
OK, NodeTraverser calls `setNode only when the visitor returns a different node. So this should be fine in most cases.

**Comment by fabpot** [comment]
New implementation with the possibility to customize the package and the version in addition to a new name (all optional).

File: src/Node/Node.php [hunk 1]
```diff
@@ -30,6 +30,10 @@ class Node implements \Countable, \IteratorAggregate
     protected $tag;
 
     private $sourceContext;
+    /** @var NameDeprecation[] */
```
Comment (stof) [inline src/Node/Node.php]: I suggest using `array<string, NameDeprecation>` to document the type of keys

Comment (fabpot) [reply src/Node/Node.php]: Done

File: src/Node/Node.php [hunk 2]
```diff
@@ -109,14 +113,37 @@ public function getAttribute(string $name)
             throw new \LogicException(\sprintf('Attribute "%s" does not exist for Node "%s".', $name, static::class));
         }
 
+        if (isset($this->attributeNameDeprecations[$name])) {
+            $dep = $this->attributeNameDeprecations[$name];
+            if ($dep->getNewName()) {
+                trigger_deprecation($dep->getPackage(), $dep->getVersion(), 'Getting attribute "%s" on a "%s" class is deprecated, get the "%s" attribute instead.', $name, static::class, $dep->getNewName());
+            } else {
+                trigger_deprecation($dep->getPackage(), $dep->getVersion(), 'Getting attribute "%s" on a "%s" class is deprecated and will throw an exception in 4.0.', $name, static::class);
```
Comment (nicolas-grekas) [inline src/Node/Node.php]: do we have to hardcode 4.0?

File: src/Node/Node.php [hunk 3]
```diff
@@ -109,14 +113,37 @@ public function getAttribute(string $name)
...
     }
 
     public function setAttribute(string $name, $value): void
     {
+        if (isset($this->attributeNameDeprecations[$name])) {
+            $dep = $this->attributeNameDeprecations[$name];
+            if ($dep->getNewName()) {
+                trigger_deprecation($dep->getPackage(), $dep->getVersion(), 'Setting attribute "%s" on a "%s" class is deprecated, set the "%s" attribute instead.', $name, static::class, $dep->getNewName());
+            } else {
+                trigger_deprecation($dep->getPackage(), $dep->getVersion(), 'Setting attribute "%s" on a "%s" class is deprecated and with no alternative in 4.0.', $name, static::class);
```
Comment (nicolas-grekas) [inline src/Node/Node.php]: 4.0?

File: src/Node/Node.php (hunk 2 above)
Comment (fabpot) [reply src/Node/Node.php]: Removed

File: src/Node/Node.php (hunk 3 above)
Comment (fabpot) [reply src/Node/Node.php]: Removed

**Review by nicolas-grekas — APPROVED** [review]

**Pull request merged**